src/main/flight/pid.c

```diff
@@ -569,6 +570,21 @@ STATIC_UNIT_TESTED FAST_CODE_NOINLINE float pidLevel(int axis, const pidProfile_
 #ifdef USE_GPS_RESCUE
     angleTarget += gpsRescueAngle[axis] / 100.0f; // Angle is in centidegrees, stepped on roll at 10Hz but not on pitch
 #endif
+#ifdef USE_POS_HOLD_MODE
+    if (FLIGHT_MODE(POS_HOLD_MODE)) {
+        angleFeedforward = 0.0f; // otherwise the lag of the PT3 carries recent stick inputs into the hold
+        if (isAutopilotActive()) {
+            // sticks are not deflected
+            angleTarget = autopilotAngle[axis]; // autopilotAngle in degrees
+            angleLimit = 85.0f; // allow autopilot to use whatever angle it needs to stop
+        }
+        // limit pilot requested angle to half the autopilot angle to avoid excess speed and chaotic stops
+        angleLimit = fminf(0.5f * apConfig()->max_angle, angleLimit);
```

Comment:
Yes it is a bit ugly.
The autopilot max angle is the max angle that autopilot code can tilt the quad to, in practice applicable only when stopping.
The allowed max angle that they can command with sticks is limited here so that the max forward speed is constrained to a value that will stop relatively easily without hitting the sanity checks and without excessively abrupt stop behaviour.
We could provide separate max Ange values for user commanded limit and for output control limit, where the control limit is set to the angle mode angleLimit, and the user's commanded limit is set to whatever they want, maybe limiting to 45 degrees.
I did it that way originally.
But then people complained it was easy to set it up to be too aggressive, and hard stops at requested max of 85 degrees can go over 90 degrees transiently and go a bit crazy.
This approach keeps everything in control.
Maybe if people complain we can look at it later.

Reply:
> Yes it is a bit ugly. The autopilot max angle is the max angle that autopilot code can tilt the quad to, in practice applicable only when stopping. 

Autopilot code already limits to maxAngle. This will apply another limiting, and to half of user specified angle (and in each angle separately - roll+pitch braking is limited less than single axis alone).
If lower angle is desired, user should enter smaller value.

> The allowed max angle that they can command with sticks is limited here so that the max forward speed is constrained to a value that will stop relatively easily without hitting the sanity checks and without excessively abrupt stop behaviour. 

But abrupt stop will happen anyway when poshold is engaged, possibly from maximum `angle_limit`. And user did specify some value in `pidProfile->angle_limit`, IMO even poshold should respect this value. Or apConfig()->max_angle may be used for poshold sticks, but then don't limit to half of user-specified value.

> We could provide separate max Ange values for user commanded limit and for output control limit, where the control limit is set to the angle mode angleLimit, and the user's commanded limit is set to whatever they want, maybe limiting to 45 degrees. I did it that way originally. But then people complained it was easy to set it up to be too aggressive, and hard stops at requested max of 85 degrees can go over 90 degrees transiently and go a bit crazy. This approach keeps everything in control. Maybe if people complain we can look at it later.

There is angle_limit and max_angle. You can limit `cli` max  for max_angle if necessary. But it is strange to allow user to set limit, but then ignore it because someone set it too high.

(BTW: maximum power braking algorithm shall be quite easy tom implement and it will fix problems with angle close to 90 deg.)

Altitude controller should be updated to limit power and high tilt angle. IMO currently it will either go to max at 90ged, or do something stupid over 90.

Reply:
Sorry I overlooked this comment in the last review. 

A suitable alternative could be to let the user set their controllable max angle and limit the range of adjustment to say 0-40 deg, then internally limit the max angle to that the control algorithm can use to twice that value. Linking the two and having only one user-configurable value makes sense to me.

Reply:
But there are 2 user-configurable values - `angle_limit` and `max_angle`

Reply:
Yes, but:
- `angle_limit` is the max that that the user can access when flying in general angle mode.  It could be set very low, or very high.  If set very low, the craft will never go very fast. If set high, it could go fast, leading to very hard stops.  However the transition from angle to autopilot is just once, when enabling it.  The transition from sticks out to sticks in can happen many times while flying around in position hold.
- position hold max_angle is ambiguous.  Do we mean the max that the control code can apply (like the old GPS Rescue code had a max)?  Or the max that the pilot can apply while flying around in Position Hold.

If we provide three configurable values:
- angle mode max
- maximum that autopilot can use 
- maximum that pilot is allowed in pos hold
... then all bases are covered.

But if possible we should avoid this.

I do know that if we constrain the max allowed pilot angle in-flight while in position hold to a relatively low value, the stops are less aggressive because the pilot cannot achieve very high speed.  This fits with the intent of stick adjustments of position and the user need for smoothness.

So the intent I had was to keep the max pilot-angle while in position hold to half the max value that the autopilot code could achieve while stopping.

This ensured that authority was sufficient in the stop for the velocity achieved.  Because if angle was limited while in position hold, usually it caused either a roll or pitch offset, and an unwelcome movement, because the limit affected only one of the two controlled axes.

Hence I think we should keep that concept active.  And hence the existing code, the intent of which is:
- pilot is limited to min of half autopilot angle or their angle mode angle (since they could have set their angle mode limit very low for some reason, and we should honour that)

II don't mind if we do the reverse where the max allowed by pilot is say 40 deg and internally we have a limit on what autopilot can do to 80 say.

Reply:
Ok .. Personally, I'd expect competent pilot and let him set angle/poshold stick limits as desired, using single config option.
Hard stop - you don't have pu push sticks all the way. And being able to stop from high speed is nice from testing - just push sticks, no need to enter/live poshold mode

> I do know that if we constrain the max allowed pilot angle in-flight while in position hold to a relatively low value, the stops are less aggressive because the pilot cannot achieve very high speed. 

Sounds like 'the algorithm has some limitations, make it harder to see them'.

> This ensured that authority was sufficient in the stop for the velocity achieved. Because if angle was limited while in position hold, usually it caused either a roll or pitch offset, and an unwelcome movement, because the limit affected only one of the two controlled axes.

Old code did this, it was fixed